build_tools/third_party/s3_management/manage.py

```diff
@@ -0,0 +1,494 @@
+
+    def _resolve_subdir(self, subdir: Optional[str] = None) -> str:
+        if not subdir:
+            subdir = self.prefix
+        # make sure we strip any trailing slashes
+        return subdir.rstrip("/")
+
+    def gen_file_list(
+        self,
+        subdir: Optional[str] = None,
+        package_name: Optional[str] = None
+    ) -> Iterable[S3Object]:
+        objects = self.objects
+        subdir = self._resolve_subdir(subdir) + '/'
+        for obj in objects:
+            if package_name is not None and self.obj_to_package_name(obj) != package_name:
+                continue
+            if self.is_obj_at_root(obj) or obj.key.startswith(subdir):
+                yield obj
+
+    def get_package_names(self, subdir: Optional[str] = None) -> List[str]:
+        return sorted({self.obj_to_package_name(obj) for obj in self.gen_file_list(subdir)})
+
+    def normalize_package_version(self: S3IndexType, obj: S3Object) -> str:
+        # removes the GPU specifier from the package name as well as
+        # unnecessary things like the file extension, architecture name, etc.
+        return sub(
+            r"%2B.*",
+            "",
+            "-".join(path.basename(obj.key).split("-")[:2])
+        )
+
+    def obj_to_package_name(self, obj: S3Object) -> str:
+        return path.basename(obj.key).split('-', 1)[0].lower()
+
+    def to_simple_package_html(
+        self,
+        subdir: Optional[str],
+        package_name: str
+    ) -> str:
+        """Generates a string that can be used as the package simple HTML index
+        """
+        out: List[str] = []
+        # Adding html header
+        out.append('<!DOCTYPE html>')
+        out.append('<html>')
+        out.append('  <body>')
+        out.append('    <h1>Links for {}</h1>'.format(package_name.lower().replace("_", "-")))
+        for obj in sorted(self.gen_file_list(subdir, package_name)):
+            # Do not include checksum for nightly packages, see
+            # https://github.com/pytorch/test-infra/pull/6307
+            maybe_fragment = f"#sha256={obj.checksum}" if obj.checksum and not obj.orig_key.startswith("whl/nightly") else ""
+            attributes = ""
+            if obj.pep658:
+                pep658_sha = f"sha256={obj.pep658}"
+                # pep714 renames the attribute to data-core-metadata
+                attributes = (
+                    f' data-dist-info-metadata="{pep658_sha}" data-core-metadata="{pep658_sha}"'
+                )
+            # Ugly hack: mark networkx-3.3, 3.4.2 as Python-3.10+ only to unblock https://github.com/pytorch/pytorch/issues/152191
+            if any(obj.key.endswith(x) for x in ("networkx-3.3-py3-none-any.whl", "networkx-3.4.2-py3-none-any.whl")):
+                attributes += ' data-requires-python="&gt;=3.10"'
+
+            out.append(
+                f'    <a href="/{obj.key}{maybe_fragment}"{attributes}>{path.basename(obj.key).replace("%2B","+")}</a><br/>'
+            )
```

Comment:
Looks like we can simplify this too (no python3.10 support, dunno about the pep658 stuff)